Implement Chiseled Bookshelf

HimmelKreis4865 opened the pull request:
## Introduction
This PR implements functionality for the Chiseled Bookshelf, added in Minecraft 1.20

### Relevant issues
Sounds are not available as I was not able to make the sound appear.
For now it seems like there is some strange extra data.

## Tests
- Block faces are supported, break time is as in vanilla.
- The Bookshelf contains up to 6 books, book & quill, enchanted books, written books
- Drops of the block are only the books stored inside, except for breaking with a tool that is enchanted with silk touch, in this case the bookshelf is dropped too
- If the players inventory is full, the book will drop (starting from the eye position, moving towards the bookshelf)
- Books can be taken with any item, no blocks can be placed in front of the books (only possible if the player is sneaking)
- Item data did not change over several server restarts

HimmelKreis4865 committed:
implement chiseled bookshelf

HimmelKreis4865 committed:
test block consistency

HimmelKreis4865 committed:
fix inconsistent operator

HimmelKreis4865 committed:
fix phpstan problems

HimmelKreis4865 committed:
fix phpstan problems #2

HimmelKreis4865 committed:
fix phpstan problems 2

HimmelKreis4865 committed:
Merge branch 'minor-next' of https://github.com/HimmelKreis4865/PocketMine-MP into minor-next

ShockedPlot7560 commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,144 @@
+	}
+
+	public function getDrops(Item $item) : array{
+		return ($item->hasEnchantment(VanillaEnchantments::SILK_TOUCH()) ? [$this->asItem(), ...$this->items] : $this->items);
```
Use ``getSilkTouchDrops`` rather than this check

HimmelKreis4865 replied:
oh yeah i actually didn't realize `BlockBreakInfo->isToolCompatible()` would return true if harvest level is 0.

HimmelKreis4865 committed:
use getSilkTouchDrops() instead of workaround

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,147 @@
+		$this->items[$index] = $item;
+	}
+
+	public function recalculateBooksBit() : int{
```
This should not be public

dktapps replied:
In addition, this is implicitly exposing Minecraft blockstate internals to the API

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,147 @@
+		return $this->booksBit = $bit;
+	}
+
+	public function setCachedBooksBit(int $booksBit) : self{
```
This is premature optimisation

HimmelKreis4865 replied:
It's only usecase is BlockToObjectDeserializer actually

dktapps replied:
Then it should be encoded in there, not here. This internal stuff shouldn't be exposed to the API, and encoding 6 bool flags doesn't require a cache.

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,147 @@
+
+	protected function describeBlockOnlyState(RuntimeDataDescriber $w) : void{
+		$w->horizontalFacing($this->facing);
+		$w->int(6, $this->booksBit);
```
This should use 6 bools, one for each occupied slot

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,147 @@
+		return $this->items[$index] ?? null;
+	}
+
+	public function setItem(int $index, WritableBookBase|Book|null $item) : void{
```
This should be fluent for consistency with other setters

HimmelKreis4865 committed:
split internals from api

HimmelKreis4865 committed:
fix missing const import

diamond-gold commented on src/block/tile/TileFactory.php:
```diff
@@ -59,6 +59,7 @@ public function __construct(){
 		$this->register(BrewingStand::class, ["BrewingStand", "minecraft:brewing_stand"]);
 		$this->register(Cauldron::class, ["Cauldron", "minecraft:cauldron"]);
 		$this->register(Chest::class, ["Chest", "minecraft:chest"]);
+		$this->register(ChiseledBookshelf::class, ["Chiseled Bookshelf", "minecraft:chiseled_bookshelf"]);
```
```suggestion
		$this->register(ChiseledBookshelf::class, ["ChiseledBookshelf", "minecraft:chiseled_bookshelf"]);
```
Vanilla ID does not contain space

HimmelKreis4865 committed:
remove invalid space

HimmelKreis4865 commented:
are all problems resolved?

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+		if($face !== $this->getFacing()){
+			return false;
+		}
+		$x = (Facing::axis($face) === Axis::X ? $clickVector->getZ() : $clickVector->getX());
```
```suggestion
		$x = Facing::axis($face) === Axis::X ? $clickVector->getZ() : $clickVector->getX();
```

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($face !== $this->getFacing()){
```
```suggestion
		if($face !== $this->facing){
```
getter is not necessary here

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+
+	public function setItem(int $index, WritableBookBase|Book|null $item) : self{
+		if($item === null){
+			if(isset($this->items[$index])){
```
useless `isset`

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+	 * @param (WritableBookBase|Book)[] $items
+	 */
+	public function setItems(array $items) : self{
+		$this->items = $items;
```
Multiple problems here:

- The number of items must be at most 6
- The indexes in the given array must not exceed 5 (`array_values()` probably OK for this)
- The array must be ordered
- The `Item`s in the array must be `clone`d

HimmelKreis4865 replied:
array_values will not work as it may skip book slots, why would the array need to be ordered?

dktapps replied:
If we foreach over it, it may be implicitly expected.

dktapps replied:
Actually, forget about the ordering thing - looks like this is a more widespread problem anyway

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+	 * @return (WritableBookBase|Book)[]
+	 */
+	public function getItems() : array{
+		return $this->items;
```
These items need to be `clone`d

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+		$x = min(match($face){
+			Facing::NORTH, Facing::EAST => 1 - $x,
+			default => $x
+		}, 0.9);
```
I feel like this variable would be more appropriately named `horizontal`.

Also, what's the deal with this 0.9?

HimmelKreis4865 replied:
0.9 is just a number between `0.66667 < x < 1` I used it so `floor($x * 3)` is scaled to be between 0-2

dktapps replied:
Seems to me that `(int) (fmod($x, 1) * 3)` is what's needed here.

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+				$this->setItem($index, null);
+			}
+		}else{
+			 if($this->getItem($index) instanceof Item){
```
```suggestion
			 if($this->getItem($index) !== null){
```

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+		}, 0.9);
+		$index = ($clickVector->y < 0.5 ? 3 : 0) + (int) floor($x * 3);
+		if(!$item instanceof WritableBookBase && !$item instanceof Book){
+			if($player instanceof Player){
```
```suggestion
			if($player !== null){
```

dktapps commented on src/block/tile/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,82 @@
+		if($itemsTag !== null && $itemsTag->getTagType() === NBT::TAG_Compound){
+			/** @var CompoundTag $itemNBT */
+			foreach($itemsTag->getValue() as $itemNBT){
+				/** @var WritableBookBase|Book $item */
```
This is a bad idea. There's no guarantee that `Item::nbtDeserialize()` will actually return a book item.

dktapps commented on src/data/bedrock/block/convert/BlockStateDeserializerHelper.php:
```diff
@@ -373,4 +378,14 @@ public static function mapWoodenSlabType(BlockStateReader $in) : Slab{
+	public static function mapChiseledBookshelf(BlockStateReader $in) : Block{
+		$bit = $in->readInt(StateNames::BOOKS_STORED);
+		$items = array_filter(array_fill(0, 6, VanillaItems::BOOK()),
+			fn(Book $book, int $slot) => (($bit & (1 << $slot)) === (1 << $slot)),
```
```suggestion
			fn(Book $book, int $slot) => ($bit >> $slot) & 1 === 1,
```

dktapps commented on src/data/bedrock/block/convert/BlockStateDeserializerHelper.php:
```diff
@@ -373,4 +378,14 @@ public static function mapWoodenSlabType(BlockStateReader $in) : Slab{
+		$bit = $in->readInt(StateNames::BOOKS_STORED);
+		$items = array_filter(array_fill(0, 6, VanillaItems::BOOK()),
+			fn(Book $book, int $slot) => (($bit & (1 << $slot)) === (1 << $slot)),
+			ARRAY_FILTER_USE_BOTH);
```
This array_* magic looks really weird and it's hard to understand what's going on. A simple `foreach` or `for` would be preferable.

dktapps commented on src/data/bedrock/block/convert/BlockStateDeserializerHelper.php:
```diff
@@ -373,4 +378,14 @@ public static function mapWoodenSlabType(BlockStateReader $in) : Slab{
 	}
+
+	public static function mapChiseledBookshelf(BlockStateReader $in) : Block{
+		$bit = $in->readInt(StateNames::BOOKS_STORED);
```
`boundedInt` should be used here.

dktapps commented on src/data/bedrock/block/convert/BlockStateDeserializerHelper.php:
```diff
@@ -373,4 +378,14 @@ public static function mapWoodenSlabType(BlockStateReader $in) : Slab{
+
+	public static function mapChiseledBookshelf(BlockStateReader $in) : Block{
+		$bit = $in->readInt(StateNames::BOOKS_STORED);
+		$items = array_filter(array_fill(0, 6, VanillaItems::BOOK()),
```
The number of books should not be hardcoded

dktapps commented on src/data/bedrock/block/convert/BlockStateSerializerHelper.php:
```diff
@@ -288,4 +290,15 @@ public static function encodeWoodenSlab(Slab $block, string $typeValue) : BlockS
 	}
+
+	public static function encodeChiseledBookshelf(ChiseledBookshelf $block) : BlockStateWriter{
+		$slots = array_keys($block->getItems());
```
This `array_keys()` calls is redundant. A `foreach($block->getItems() as $slot => $item)` can be used instead.

dktapps commented on src/data/bedrock/block/convert/BlockStateSerializerHelper.php:
```diff
@@ -288,4 +290,15 @@ public static function encodeWoodenSlab(Slab $block, string $typeValue) : BlockS
+
+	public static function encodeChiseledBookshelf(ChiseledBookshelf $block) : BlockStateWriter{
+		$slots = array_keys($block->getItems());
+		$bit = 0;
```
`flags` would be a more idiomatic name.

dktapps commented on src/block/tile/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,82 @@
+use function array_keys;
+use function array_map;
+
+class ChiseledBookshelf extends Spawnable{
```
Is it necessary for this to be `Spawnable`? Since we're not sending any special NBT data anyway, it ought to work fine with just a regular `Tile` as far as I can tell.

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+				$item = $this->getItem($index) ?? VanillaItems::AIR();
+
+				$leftover = $player->getInventory()->addItem($item);
+				if(count($leftover) > 0){
```
The common method is to use a `foreach` for this. However, this entire section could likely be avoided by using `returnedItems` (which would also not require the `player`).

dktapps commented on src/block/ChiseledBookshelf.php:
```diff
@@ -0,0 +1,145 @@
+			 if($this->getItem($index) instanceof Item){
+				 return false;
+			 }
+			 /** @var WritableBookBase|Book $item */
```
Properly written code should not be using these within internals

dktapps committed:
Merge branch 'minor-next' into chiseled-bookshelf

dktapps committed:
...

dktapps committed:
Tidy

dktapps committed:
Clean up serializers

dktapps committed:
Use the same model as brewing stand for slot flags
instead of hard-linking them to the inventory contents, we allow the flags to simply represent the _appearance_ of the block, rather than its actual contents.
This allows a range of new use cases, and is also consistent with brewing stands, which behave in the same way.

dktapps committed:
Allow enchanted books to be put in a chiselled bookshelf

dktapps committed:
Chiseled Bookshelf tile is not required to be spawnable
since it has no spawn data and no rendering effect, it's unnecessary to send it on the network.

dktapps committed:
Remove unnecessary getter usage

dktapps committed:
Remove unnecessary helpers
these are only used once
